protocols/identify/src/protocol.rs

```diff
@@ -184,77 +188,99 @@ where
         protocols: info.protocols,
     };
 
-    let mut bytes = Vec::with_capacity(message.encoded_len());
-    message
-        .encode(&mut bytes)
-        .expect("Vec<u8> provides capacity as needed");
+    let mut framed_io = FramedWrite::new(
+        io,
+        prost_codec::Codec::<structs_proto::Identify>::new(MAX_MESSAGE_SIZE_BYTES),
+    );
 
-    upgrade::write_length_prefixed(&mut io, bytes).await?;
-    io.close().await?;
+    framed_io.send(message).await?;
+    framed_io.close().await?;
 
     Ok(())
 }
 
-async fn recv<T>(mut socket: T) -> io::Result<IdentifyInfo>
+async fn recv<T>(mut socket: T) -> Result<IdentifyInfo, UpgradeError>
 where
     T: AsyncRead + AsyncWrite + Unpin,
 {
     socket.close().await?;
```

Comment:
Just to help my understanding: Why do we close the socket here before creating the `FramedRead`, but don't do this in dcutr on `upgrade_inbound`?

Reply:
The standard identify exchange (not identify push) opens a stream to a remote, never writes on it, and expects the remote to send the response on said stream. Given that the local node never writes on the stream it opened it is safe to close it early. Not sure it is strictly necessary though.

(Also given that the local node never writes on the channel, it can use a `FramedRead` instead of a `Framed`.)

In DCUtR `upgrade_inbound` the remote node sends a request and the local node replies with a response. Closing the stream before creating the `Framed` would prevent the local node from being able to respond.

Does the above explanation make sense @elenaf9?

Reply:
Yes makes sense, thanks!